Домашнее задание

RushianHaker opened the pull request:
**Домашнее задание:** Создать приложение хранящее информацию о книгах в библиотеке
**Цель:** использовать возможности Spring JDBC и spring-boot-starter-jdbc для подключения к реляционным базам данных
**Результат:** приложение с хранением данных в реляционной БД, которое в дальнейшем будем развивать

helloDearSupportTeam committed:
init

RushianHaker committed:
Update spring-05 README.md

helloDearSupportTeam committed:
Merge branch 'main' into feature/test_for_students_5

helloDearSupportTeam committed:
чистка и новая структура

helloDearSupportTeam committed:
фикс мапперов

helloDearSupportTeam committed:
фикс скриптов

helloDearSupportTeam committed:
фикс метода update

helloDearSupportTeam committed:
мелкие правки скриптов sql и моделей, а также отображения в консоли теста

helloDearSupportTeam committed:
test config

helloDearSupportTeam committed:
фикс тест скриптов

helloDearSupportTeam committed:
правка тестов

helloDearSupportTeam committed:
правка тестов и todo

helloDearSupportTeam committed:
@Rollback

helloDearSupportTeam committed:
clean and add field author_year

helloDearSupportTeam committed:
готовая версия

helloDearSupportTeam committed:
+ spring shell

helloDearSupportTeam committed:
+ spring shell

helloDearSupportTeam committed:
check style

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/dao/impl/AuthorDaoJdbcImpl.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Override
+    public void create(Author author) {
+        namedParameterJdbcOperations.update("insert into authors (id, \"name\", \"year\") values (:id, :name, :year)",
```
Откуда возьмется id?

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/dao/impl/AuthorDaoJdbcImpl.java:
```diff
@@ -0,0 +1,55 @@
+    @Override
+    public void update(Author author, long id) {
+        namedParameterJdbcOperations.update(
+                "update authors set id = :id, \"name\" = :name, \"year\" = :year where id = :search_id",
```
Зачем тут перезаписывать id? Он же не меняется

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/service/impl/BookServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+
+    private final IOService ioService;
+
+    private final UserAnswerServiceImpl userAnswerService;
```
Лучше по интерфейсу

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/service/impl/BookServiceImpl.java:
```diff
@@ -0,0 +1,101 @@
+    public String create() {
+        ioService.outputString("Enter books info, please:");
+
+        var bookId = userAnswerService.checkUserAnswer("- Enter book id: ");
```
Пользователь никак не может угадать с id. К БД может быть подключено куча клиентов. Лучше пусть id генерируется на стороне базы

stvort replied:
Ну и в сущностях id это `long`, а читается `int`

stvort commented on spring-05/testing-students/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,25 @@
+    id          BIGINT PRIMARY KEY,
+    "name"      VARCHAR(255),
+    "year"      int4,
+    author      VARCHAR(255),
+    genre       VARCHAR(255),
```
Лучше внешний ключ на таблицы авторов и жанров

stvort commented on spring-05/testing-students/src/test/java/ru/otus/testing/dao/DaoTestSpringBootConfiguration.java:
```diff
@@ -0,0 +1,10 @@
+
+@ComponentScan(value = "ru.otus.testing.dao")
+@SpringBootConfiguration
+public class DaoTestSpringBootConfiguration {
```
Такие штуки не нужны. Достаточно просто импортировать нужное дао в тест, как это было в примерах к занятию

stvort commented on spring-05/testing-students/src/test/java/ru/otus/testing/service/ServiceTestSpringBootConfiguration.java:
```diff
@@ -0,0 +1,10 @@
+
+@ComponentScan(value = "ru.otus.testing.service")
+@SpringBootConfiguration
+public class ServiceTestSpringBootConfiguration {
```
Эта конфигурация тоже избыточна

Commit:
почти закончил исправлять

helloDearSupportTeam committed:
фикс теста репозитория

helloDearSupportTeam committed:
фикс теста сервиса

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/TestingStudentsApplication.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class TestingStudentsApplication {
```
Почему тестирование студентов?

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/model/Author.java:
```diff
@@ -0,0 +1,25 @@
+
+    private final String name;
+
+    private final Long year;
```
Если не может быть `null`, то лучше примитив

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/service/impl/BookServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+        var genreName = ioService.readNextWithPrompt("- Enter books genre name: ");
+
+        //получение id's из таблицы books
+        var ids = bookDao.getByIdAuthorAndGenreIds(bookId);
```
Не стоит так делать. Мапы вообще не очень хорошо возвращать/принимать в публичные методы. Лучше всю книгу загрузить, поменять и сохранить

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/service/impl/BookServiceImpl.java:
```diff
@@ -0,0 +1,119 @@
+
+        var bookId = userAnswerService.checkUserAnswer("- Enter book id: ");
+
+        return bookDao.getById(bookId).toString();
```
Мы договаривались не использовать `toString()` для вывода пользователю

stvort commented on spring-05/testing-students/src/test/resources/schema.sql:
```diff
@@ -0,0 +1,31 @@
+DROP TABLE IF EXISTS authors;
```
Схема должна быть одна для тестов и для приложения. А данные разные

stvort replied:
> Схема должна быть одна для тестов и для приложения. А данные разные

Не исправлено. У тестов все еще своя схема

helloDearSupportTeam committed:
нейминг

helloDearSupportTeam committed:
привел к примитиву

helloDearSupportTeam committed:
одинаковые схемы

helloDearSupportTeam committed:
Мы договаривались не использовать toString() для вывода пользователю

helloDearSupportTeam committed:
Не стоит так делать. Мапы вообще не очень хорошо возвращать/принимать в публичные методы. Лучше всю книгу загрузить, поменять и сохранить

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/service/impl/BookServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+        this.userAnswerService = userAnswerService;
+    }
+
+    @Transactional
+    @Override
+    public Book create() {
+        ioService.outputString("Enter books info, please:");
+
+        var bookName = ioService.readNextWithPrompt("- Enter book name: ");
+        var bookYear = userAnswerService.checkUserAnswer("- Enter book written year: ");
+
+        var authorName = ioService.readNextWithPrompt("- Enter author name: ");
+        var authorYear = userAnswerService.checkUserAnswer("- Enter author years: ");
```
Видимо не заметил в прошлый раз транзакцию. Внутри нее лучше вводом-выводом не заниматься. Ну, по крайней мере, ожидать ввода пользователя

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/service/impl/BookServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+
+        var genreName = ioService.readNextWithPrompt("- Enter books genre name: ");
+
+        Genre genre = genreDao.create(new Genre(genreName));
+        Author author = authorDao.create(new Author(authorName, authorYear));
+        return bookDao.create(new Book(bookName, bookYear, author, genre));
```
Как можно создать книгу с существующим автором и/или жанром?

RushianHaker replied:
Добрый вечер, почему нельзя ?

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/service/impl/BookServiceImpl.java:
```diff
@@ -0,0 +1,122 @@
+        var genreName = ioService.readNextWithPrompt("- Enter books genre name: ");
+
+        //получение id's из таблицы books
+        var bookDaoById = bookDao.getById(bookId);
```
Это не дао же? Это же книга

stvort commented on spring-05/testing-students/src/test/java/ru/otus/testing/dao/impl/BookDaoJdbcImplTest.java:
```diff
@@ -0,0 +1,112 @@
+
+    @Test
+    void create() {
+        template.update("insert into authors (authors_name, author_year) values ('Tolstoy', 50);");
+        template.update("insert into genres (genres_name) values ('history');");
```
Лучше через data.sql подготовить данные

helloDearSupportTeam committed:
Не исправлено. У тестов все еще своя схема

helloDearSupportTeam committed:
Лучше через data.sql подготовить данные

helloDearSupportTeam committed:
Это не дао же? Это же книга

helloDearSupportTeam committed:
Видимо не заметил в прошлый раз транзакцию. Внутри нее лучше вводом-выводом не заниматься. Ну, по крайней мере, ожидать ввода пользователя

helloDearSupportTeam committed:
Как можно создать книгу с существующим автором и/или жанром?
+ поправил обновление книги, теперь обновляется только сама книга с проверкой существования жанра/автора

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/commandline/BookCommandLine.java:
```diff
@@ -0,0 +1,76 @@
+import ru.otus.testing.service.UserAnswerService;
+
+@ShellComponent
+public class BookCommandLine {
```
Это не командная строка книги

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/commandline/BookCommandLine.java:
```diff
@@ -0,0 +1,76 @@
+    public String create() {
+        ioService.outputString("Enter books info, please:");
+
+        var bookName = ioService.readNextWithPrompt("- Enter book name: ");
+        var bookYear = userAnswerService.checkUserAnswer("- Enter book written year: ");
+
+        var authorName = ioService.readNextWithPrompt("- Enter author name: ");
+        var authorYear = userAnswerService.checkUserAnswer("- Enter author years: ");
+
+        var genreName = ioService.readNextWithPrompt("- Enter books genre name: ");
```
По идее, для этого и нужны аргументы команд

RushianHaker replied:
я решил явно оставить вывод этих полей, для своего рода более явного диалога с пользоваетелем

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/service/impl/BookServiceImpl.java:
```diff
@@ -0,0 +1,114 @@
+    @Override
+    public Book readById() {
+        ioService.outputString("Enter books id, that info you want to see: ");
+        var bookId = userAnswerService.checkUserAnswer("- Enter book id: ");
```
Так или иначе, этим лучше заниматься на одном уровне. А то странно. Для `create` мы данные извне получаем, а для `readById` внутри

stvort commented on spring-05/testing-students/src/main/java/ru/otus/testing/service/impl/BookServiceImpl.java:
```diff
@@ -0,0 +1,114 @@
+        var bookId = userAnswerService.checkUserAnswer("- Enter book id: ");
+
+        var bookInfo = bookDao.getById(bookId);
+        ioService.outputString(
+                "Book-" + bookInfo.getId() + ")" +
+                        " id: " + bookInfo.getId() +
+                        ", name: " + bookInfo.getName() +
+                        ", year: " + bookInfo.getYear() +
+                        ", author: " + bookInfo.getAuthor().getName() +
+                        ", author years: " + bookInfo.getAuthor().getYear() +
+                        ", genre: " + bookInfo.getGenre().getName());
```
С выводом тоже самое. Причем конвертеры в строку лучше сделать отдельно. Примерно такая схема получается:
Команда:
1. принимет от пользователя аргументы
2. вызывает сервис
3. вызывает конвертер для преобразования результата сервиса в строку
4. возвращает эту строку

Сервис ввода-вывода можно выкидывать при такой схеме

helloDearSupportTeam committed:
Это не командная строка книги

helloDearSupportTeam committed:
Так или иначе, этим лучше заниматься на одном уровне. А то странно. Для create мы данные извне получаем, а для readById внутри

helloDearSupportTeam committed:
clean

helloDearSupportTeam committed:
нейминг

helloDearSupportTeam committed:
нейминг
